docs/developer-docs/smart-contracts/advanced-features/async-code.mdx

```diff
@@ -1,90 +1,94 @@
 
-In the following example, another call is added to a local function to canister A before making the inter-canister call to canister B.
+## Guaranteed response vs best-effort response calls
+
+ICP supports two kind of inter-canister calls:
+
+* Guaranteed response calls provide the caller with the guarantee that, if the callee produces a response (which may be unsuccessful, i.e., an error during call processing), that exact response will be delivered to the caller. Furthermore, if the request isn't successfully delivered to the callee (which can happen during high load, callee running out of cycles, and other reasons), the response will notify the caller of this.
+
+* Best-effort response calls drop this guarantee. They allow the caller to specify a timeout (which is capped from above by the system to some maximum value, such as 5 minutes), after which the system will return an error to the caller. The request may or may not be processed by the callee; the caller must, if necessary, determine whether the call took place or not by some other mechanism.
+
+Guaranteed response calls are a legacy call mechanism. While on the surface they make it easier to write correct applications, they also suffer some significant drawbacks:
+
+* If the callee is unresponsive (which could happen because of high load on the callee, high load or an error on the callee's subnet, or even because the callee is malicious and delays the response on purpose), the caller canister is stalled and has no control over when it can resume processing or provide an answer. Furthermore, the caller canister cannot be safely stopped and upgraded.
+* They do not scale well. When the system is under high load, canisters may be unable to issue further inter-canister calls.
+
+Best-effort calls address both of these problems. However, they can require more effort on the part of the caller to handle the additional error case. Here are some guidelines how to choose between the two types of calls:
+
+* Always prefer best-effort response calls for calls that don't change the state of the callee, i.e., reads.
+* For endpoints that change the state of the callee, the best practice is to make such endpoints amenable to [safe retries](/docs/current/developer-docs/smart-contracts/best-practices/idempotency), since this is also required for safely handling external user calls. Use best-effort responses calls for such endpoints, and handle the additional edge cases.
+* Use guaranteed response calls for endpoints that mutate state and do not enable safe retries, or calls that perform larger cycle transfers. Best-effort response calls can currently lose cycles in edge cases. Be aware of the limitations of guaranteed response calls listed above. If safe upgrades are needed, consider using a stateless proxy canister (see the [security best practices](https://internetcomputer.org/docs/current/developer-docs/security/security-best-practices/inter-canister-calls) for more information).
```

Comment:
Here I would not say that best effort calls can lose cycles in edge cases as it makes it sound as if it was a bug in the system. I'd rather use the formulation that if a message is dropped also the attached cycles will be dropped.

Reply:
Good idea, changed.